REGR: avoid bbox check in overlay if how != intersection

martinfleis opened the pull request:
Fixes #2156

The bbox check makes sense only for `intersection`, not the other overlay options.

martinfleis commented:
Okay, I know why the order behaves like that. Or at least where that happens.

This snippet is non-deterministic. I get `['col1', 'col2']` in one env with Python 3.9.6 and `['col2', 'col1']` in another. 
```
list(set(['col1']+ ['col2']))
```

Shall we ensure we return the same order of columns every time with overlay?

jorisvandenbossche commented on geopandas/tools/overlay.py:
```diff
@@ -265,23 +265,24 @@ def overlay(df1, df2, how="intersection", keep_geom_type=None, make_valid=True):
+            return GeoDataFrame(
+                [],
+                columns=list(
+                    set(
+                        df1.drop(df1.geometry.name, axis=1).columns.to_list()
+                        + df2.drop(df2.geometry.name, axis=1).columns.to_list()
+                    )
```
This whole line of code seems a bit fishy: 1) I am not sure the `set(..)` is actually needed? (the only reason I can think of that is was added is to remove duplicate columns, but a non-empty overlay also doesn't do that .. (since pd.merge preserves duplicate columns)), and 2) this actually also ignores the fact that we add suffixes in the normal code path with merge if there are name clashes between left / right.

martinfleis commented:
I've removed sets and mirrored merge we use in an overlapping case. It should now return exactly the same columns in all cases. I've also changed affected (previously wrong) tests, so it should also be covered by CI.

jorisvandenbossche commented on geopandas/tools/overlay.py:
```diff
@@ -59,11 +59,15 @@ def _overlay_intersection(df1, df2):
-            [],
-            columns=list(set(df1.columns).union(df2.columns)) + ["__idx1", "__idx2"],
-            crs=df1.crs,
+        result = df1[:0].merge(
+            df2[:0].drop(df2.geometry.name, axis=1),
```
```suggestion
        result = df1.iloc[:0].merge(
            df2.iloc[:0].drop(df2.geometry.name, axis=1),
```

martinfleis commented on geopandas/tests/test_overlay.py:
```diff
@@ -656,5 +657,69 @@ def test_empty_overlay_return_non_duplicated_columns():
+                "geometry": [],
+            },
+            index=pd.Index([], dtype="object"),
+        )
```
I figured this out as well. I just came to a conclusion that `check_dtype=False` will do in this case. Happy to keep it though!
